yarn-project/native/src/msgpack_channel.ts

```diff
@@ -0,0 +1,109 @@
+import { Fr } from '@aztec/foundation/fields';
+import { MessageHeader, TypedMessage } from '@aztec/foundation/message';
+
+import { Encoder, addExtension } from 'msgpackr';
+import { isAnyArrayBuffer } from 'util/types';
+
+export interface MessageReceiver {
+  call(msg: Buffer | Uint8Array): Promise<Buffer | Uint8Array>;
+}
+
+export type RoundtripDuration = {
+  encodingUs: number;
+  callUs: number;
+  decodingUs: number;
+  totalUs: number;
+};
+
+// small extension to pack an NodeJS Fr instance to a representation that the C++ code can understand
+// this only works for writes. Unpacking from C++ can't create Fr instances because the data is passed
+// as raw, untagged, buffers. On the NodeJS side we don't know what the buffer represents
+// Adding a tag would be a solution, but it would have to be done on both sides and it's unclear where else
+// C++ fr instances are sent/received/stored.
+addExtension({
+  Class: Fr,
+  write: fr => fr.toBuffer(),
+});
```

Comment:
Not gonna lie, I see this and I get scared about multiple `Fr` classes floating around, whether it's due to a different package version being loaded, or simple because we have other `Fr`s such as the one coming out from bbjs (see `barretenberg/ts/src/types/fields.ts`).